src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java

```diff
@@ -1458,13 +1469,35 @@ public void finish() throws IOException {
          * @param    force    flush data regardless whether the buffer is full
          */
         public void exitSegmentMode() throws IOException {
+            if (unbufferedMode) {
+                // no data in internal buffer.
+                assert segmentWritten == 0;
+                unbufferedMode = false;
+            }
             if (allowSegmented && segmentMode && shouldFlush()) {
                 flush();
                 assert segmentWritten == 0;
                 segmentMode = false;
             }
         }
 
+        /**
+         * Fill segment size and enable unbuffered mode
+         * @param    size    size of data to be written
+         */
+        public void fillSegmentSizeAndEnableUnbufferedMode(int size) throws IOException {
+            assert segmentMode == true;
+            assert unbufferedMode == false;
+            if (segmentWritten != 0) {
+                // flush previous written data and clear the internal buffer.
+                flush();
+            }
+            // buffer must be empty now.
+            assert (segmentMode && (segmentWritten == 0) && (unbufferedMode == false)) : "Wrong Status";
```

Comment:
This assert repeats checks in asserts at lines L1489 and L1490.
It is not clear why this duplication is needed.